### PR TITLE
Deprecate `mf_flag` call signature in xclim.ensembles, expose calendar conversion utilities, fix _gen_returns_section bug

### LINKED ISSUE #1305: [PyOpenSci] Reviewer #1 Comments
### Generic Issue

* xclim version: 0.40.0
* Python version: Any
* Operating System: All

### Description

There are some issues with our documentation that should be addressed. (see: https://github.com/pyOpenSci/software-submission/issues/73#issuecomment-1445193038)

The following comments received from first PyOpenSci reviewer, @aguspesce (Thank you!)

-----

#### Review Comments

Thank you for the opportunity to review this code. The package looks very good and a lot of effort has been put into its development. I only have minor suggestions and observations, so I will enumerate them in order of priority.

**About the documentation**

I found some issues on the documentation website that don't look well.

* [x] The type annotations is a great tool, but sometimes it might confuse users when they read the API documentation.
      For example, when a function has a lot of parameters, the type annotation can clutter its documentation.
      So I suggest using the `sphinx_autodoc_typehints` extension in `docs/conf.py` which hide the type annotations when sphinx creates the documentation. (#1308)

* [ ] The documentation for indicators and indices has a format problem when it describes the return part.
      For example, `xclim.indicators.atmos.cffwis_indices` has multiple **descriptions** and this makes the return part difficult to understand:
      **Returns**   dc (DataArray) – Drought Code (drought_code), with additional attributes:
      **description:** Numeric rating of the average moisture content of deep,
      compact organic layers.dmc : DataArray Duff Moisture Code (duff_moisture_code),
      with additional attributes: **description:** Numeric rating of the average
      moisture content of loosely compacted organic layers of moderate depth.ffmc :
      DataArray Fine Fuel Moisture Code (fine_fuel_moisture_code), with additional
      attributes: **description:** Numeric rating of the average moisture content
      of litter and other cured fine fuels.isi : DataArray Initial Spread Index
      (initial_spread_index), with additional attributes: **description:** Numeric
      rating of the expected rate of fire spread.bui : DataArray Buildup Index
      (buildup_index), with additional attributes: **description:** Numeric rating
      of the total amount of fuel available for combustion.fwi : DataArray Fire
      Weather Index (fire_weather_index), with additional attributes: **description:**
      Numeric rating of fire intensity. (#1317)

* [x] The API documentation is a very long page, and it is difficult to navigate and find definitions. This can be improved by adding a table of contents. (#1308)

* [x] Why does `xclim.sdba` API have a separate section, and it is not in the USER API session?  As a user, who wants a description of any functionality of this subpackage, I would go to find it in that section. I think that it's better if all the APIs are in the USER API section.(#1308)

**About the repository**

* [x] In my opinion, I like to see the licence type in the README file likes: (#1308)
      ```
      ## License
      This is free software: you can redistribute it and/or modify it under the terms
      of the Apache License 2.0.
      A copy of this license is provided in LICENSE.txt.
      ```

* [x] The `CONTRIBUTING.rst` file is in `.github/` directory.  I suggest moving it to the main directory `./` because it's easier to find it. (#1308)

* [x] When I ran the test using `make test`, only one test failed: (will be resolved in #1303)
      ```
      FAILED xclim/testing/tests/test_atmos.py::TestMeanRadiantTemperature::test_mean_radiant_temperature
       ===== 1 failed, 1398 passed, 66 skipped, 2 xfailed, 1 xpassed, 95 warnings in 315.63s (0:05:15) ====== 
      make: *** [Makefile:68: test] Error 1
      ```

**The JOSS paper**

* [x] In line number 49 you write: _"The development of `xclim` started at Ouranos in 2018 out of the need to deliver data for a pan-Canadian atlas of climate indicator."_ . What it is Oranos? Maybe a line about this or a link to the website.

**The code**

About the code, I don't have any objection. I have only noted that the code is sometimes difficult to follow. So, I just have a few observations that might help to improve this in future developments.

* [ ] For example, `xclim.ensembles.create_ensemble` has an argument called  `mf_flag`. This argument is a bool for definition, so the user can understand that it's a flag. Therefore, this argument can be called `multifile` to be more readable. (#1317)

* [ ] Other example is `xclim.core.convert_calendar`.  It has 2 private functions (`_yearly_interp_doy`, `_yearly_random_doy`) that could be out of the main function.  This makes the if statements easier to follow due to the code has a very large if statement. (#1317)

**Documentation Checklist**
 
- [x] If applicable, how the package compares to other similar packages and/or how it relates to other packages in the scientific ecosystem. (#1308)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Zeitsperre**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #1305 
- [x] Tests for the changes have been added (for bug fixes / features)
  - [x] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] CHANGES.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

* Changes the call signature for `xclim.ensembles.create_ensemble()` and `_ens_align_datasets()` to use `multifile` in lieu of `mf_flag`.
* Updates docstrings and emits warnings for deprecated usage on those call signatures.
* Exposes two previously private functions that can return day of year indexes explicitly or randomly.
* Moves the module docstring from `xclim.indices` into the indices documentation file.
* Fixes a formatting bug in `_gen_returns_section` that caused poor rendering in the documentation

### Does this PR introduce a breaking change?

No*. Original calls with kwargs `mf_flag` set will still work, but will emit warnings that users should update their scripts. This will be a major breaking change in xclim v0.43.0. The new functions were previously inaccessible (not breaking).

### CODE REVIEW

**Comment by Zeitsperre** [comment]
Question: Should I deal with the other API-related comment in this PR?

> Other example is `xclim.core.convert_calendar`. It has 2 private functions (`_yearly_interp_doy`, `_yearly_random_doy`) that could be out of the main function. This makes the if statements easier to follow due to the code has a very large if statement.

**Review by bzah — COMMENTED** [review]
LGTM! 
There are not too many changes here, so from a reviewer pov I would not mind if you also included the other API changes in the same PR.

File: xclim/ensembles/_base.py [hunk 1]
```diff
@@ -103,9 +105,18 @@ def create_ensemble(
             "is a glob pattern, as the final order is random."
         )
 
+    if mf_flag != "UNSET":
+        warnings.warn(
+            "The `mf_flag` argument is being deprecated in favour of `multifile` in `create.ensemble()`. "
+            "This change will be made effective from `xclim>=0.43.0`. Please update your scripts accordingly",
+            DeprecationWarning,
```
Comment (aulemahal) [inline xclim/ensembles/_base.py:112]:
```suggestion
            FutureWarning,
```

File: xclim/ensembles/_base.py [hunk 2]
```diff
@@ -375,10 +387,19 @@ def _ens_align_datasets(
     if isinstance(datasets, str):
         datasets = glob(datasets)
 
+    if mf_flag != "UNSET":
+        warnings.warn(
+            "The `mf_flag` argument is being deprecated in favour of `multifile` in `_ens_align_datasets()`. "
+            "This change will be made effective from `xclim>=0.43.0`. Please update your scripts accordingly",
+            DeprecationWarning,
```
Comment (aulemahal) [inline xclim/ensembles/_base.py:394]:
```suggestion
            FutureWarning,
```
Future Warnings are visible to the user, while Deprecation Warnings are not. It confuses me everythime... https://docs.python.org/3/library/warnings.html?highlight=warnings#warning-categories

File: xclim/testing/tests/test_ensembles.py [hunk 3]
```diff
@@ -244,7 +243,8 @@ def test_calc_mean_std_min_max(self, ensemble_dataset_objects, open_dataset):
         for n in ensemble_dataset_objects["nc_files_simple"]:
             ds = open_dataset(n)
             ds_all.append(ds)
-        ens = ensembles.create_ensemble(ds_all)
+        with pytest.warns(DeprecationWarning):
```
Comment (aulemahal) [inline xclim/testing/tests/test_ensembles.py:246]:
```suggestion
        with pytest.warns(FutureWarning):
```

**Review by aulemahal — APPROVED** [review]
Agreed!
We had the same Future vs Deprecation issue in xscen this morning...

File: xclim/ensembles/_base.py (hunk 2 above)
Comment (Zeitsperre) [reply xclim/ensembles/_base.py:394]: Dang. That's good to know!

**Comment by Zeitsperre** [comment]
Returns formatting for Indicators is improved:
image

**Pull request merged**